[WebRTC] Crash in webrtc::field_trial::FindFullName() calling strlen() on freed C-string

ddkilzer opened the pull request:
#### ca00a6a0f5f924657977487bd5198ddd3d833c0b
<pre>
[WebRTC] Crash in webrtc::field_trial::FindFullName() calling strlen() on freed C-string
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=281021">https://bugs.webkit.org/show_bug.cgi?id=281021</a>&gt;
&lt;<a href="https://rdar.apple.com/137463467">rdar://137463467</a>&gt;

Reviewed by Chris Dumez.

* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint):
- webrtc::field_trial::InitFieldTrialsFromString() keeps a reference to
  the C-string passed in, but it does not make its own copy of it, so
  either a global string or a &quot;leaked&quot; copy must be passed in.  This
  changes the LibWebRTCMediaEndpoint() constructor to create two global
  strings and to use whichever one is needed.

Canonical link: <a href="https://commits.webkit.org/284806@main">https://commits.webkit.org/284806@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7499ab8ab01f13f98f40470e069f49babac32b4c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70586 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49992 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23351 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74675 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21764 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72703 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57792 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21604 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/14380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73652 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60846 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18282 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20125 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18637 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76395 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14812 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17853 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63638 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60914 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/63574 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11620 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/5266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10812 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45794 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46868 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48145 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46610 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
```diff
@@ -82,11 +82,13 @@ LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint(LibWebRTCPeerConnectionBackend& p
 
-    String fieldTrials = "WebRTC-Video-H26xPacketBuffer/Enabled/"_s;
+#define WEBKIT_WEBRTC_FIELD_TRIALS "WebRTC-Video-H26xPacketBuffer/Enabled/"
+    const char* fieldTrials = WEBKIT_WEBRTC_FIELD_TRIALS;
```
We should use ASCIILiteral in new code.

ddkilzer replied:
I'm not sure that I understand.  `webrtc::field_trial::InitFieldTrialsFromString()` below takes a `const char*`, so how is using an `WTF::ASCIILiteral` going to help us here?  We're going to create the C++ wrapper object, then pull out the `const char*` again to pass it to the function...?

cdumez replied:
Just trying to follow our new safer C++ guidelines:
- https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#use-asciiliteral-instead-of-const-char-for-string-literals

The benefits in this particular instance are likely very small but it is good practice in general.

ddkilzer replied:
Okay, I have a fix.

cdumez commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
```diff
@@ -82,11 +82,13 @@ LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint(LibWebRTCPeerConnectionBackend& p
 
-    String fieldTrials = "WebRTC-Video-H26xPacketBuffer/Enabled/"_s;
+#define WEBKIT_WEBRTC_FIELD_TRIALS "WebRTC-Video-H26xPacketBuffer/Enabled/"
+    ASCIILiteral fieldTrials = WTF_CONCAT(WEBKIT_WEBRTC_FIELD_TRIALS, _s);
```
nit: We can probably use `auto`?

cdumez reviewed (approved):
LGTM

Commit:
[WebRTC] Crash in webrtc::field_trial::FindFullName() calling strlen() on freed C-string
<https://bugs.webkit.org/show_bug.cgi?id=281021>
<rdar://137463467>

Reviewed by Chris Dumez.

* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCMediaEndpoint.cpp:
(WebCore::LibWebRTCMediaEndpoint::LibWebRTCMediaEndpoint):
- webrtc::field_trial::InitFieldTrialsFromString() keeps a reference to
  the C-string passed in, but it does not make its own copy of it, so
  either a global string or a "leaked" copy must be passed in.  This
  changes the LibWebRTCMediaEndpoint() constructor to create two global
  strings and to use whichever one is needed.

Canonical link: https://commits.webkit.org/284806@main
